orca/producer_test.go

```diff
@@ -0,0 +1,538 @@
+type testOOBListener struct {
+	cleanup      func()
+	loadReportCh chan *v3orcapb.OrcaLoadReport
+}
+
+func newTestOOBListener() *testOOBListener {
+	return &testOOBListener{cleanup: func() {}, loadReportCh: make(chan *v3orcapb.OrcaLoadReport)}
+}
+
+func (t *testOOBListener) Stop() { t.cleanup() }
+
+func (t *testOOBListener) OnLoadReport(r *v3orcapb.OrcaLoadReport) {
+	t.loadReportCh <- r
+}
+
+// TestProducer is a basic, end-to-end style test of an LB policy with an
+// OOBListener communicating with a server with an ORCA service.
+func (s) TestProducer(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Use a fixed backoff for stream recreation.
+	oldBackoff := internal.DefaultBackoffFunc
+	internal.DefaultBackoffFunc = func(int) time.Duration { return 10 * time.Millisecond }
+	defer func() { internal.DefaultBackoffFunc = oldBackoff }()
+
+	// Initialize listener for our ORCA server.
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Register the OpenRCAService with a very short metrics reporting interval.
+	const shortReportingInterval = 50 * time.Millisecond
+	opts := orca.ServiceOptions{MinReportingInterval: shortReportingInterval}
+	internal.AllowAnyMinReportingInterval.(func(*orca.ServiceOptions))(&opts)
+	s := grpc.NewServer()
+	orcaSrv, err := orca.Register(s, opts)
+	if err != nil {
+		t.Fatalf("orca.Register failed: %v", err)
+	}
+	go s.Serve(lis)
+	defer s.Stop()
+
+	// Create our client with an OOB listener in the LB policy it selects.
+	r := manual.NewBuilderWithScheme("whatever")
+	oobLis := newTestOOBListener()
+
+	lisOpts := orca.OOBListenerOptions{ReportInterval: 50 * time.Millisecond}
+	lao := &listenerAndOptions{listener: oobLis, opts: lisOpts}
+	r.InitialState(resolver.State{Addresses: []resolver.Address{setListenerAndOptions(resolver.Address{Addr: lis.Addr().String()}, lao)}})
+	cc, err := grpc.Dial("whatever:///whatever", grpc.WithDefaultServiceConfig(`{"loadBalancingConfig": [{"customLB":{}}]}`), grpc.WithResolvers(r), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("grpc.Dial failed: %v", err)
+	}
+	defer cc.Close()
+
+	// Ensure the OOB listener is stopped before the client is closed to avoid
+	// a potential irrelevant error in the logs.
+	defer oobLis.Stop()
+
+	// Set a few metrics and wait for them on the client side.
+	orcaSrv.SetCPUUtilization(10)
+	orcaSrv.SetMemoryUtilization(100)
+	orcaSrv.SetUtilization("bob", 555)
+	loadReportWant := &v3orcapb.OrcaLoadReport{
+		CpuUtilization: 10,
+		MemUtilization: 100,
+		Utilization:    map[string]float64{"bob": 555},
+	}
+
+testReport:
+	for {
+		select {
+		case r := <-oobLis.loadReportCh:
+			t.Log("Load report received: ", r)
+			if proto.Equal(r, loadReportWant) {
+				// Success!
+				break testReport
+			}
+		case <-ctx.Done():
+			t.Fatalf("timed out waiting for load report of: %v", loadReportWant)
```

Comment:
load report of? s/ to load report? load report of (specific exact load report here) feels off.

Reply:
I just don't like the word of. Just delete of, and timed out waiting for load report: (specific exact load report here)

Reply:
oh "s/ to" is "change to"?  Usually we'd write "s/<old>/<new>/" not "s/ to"